idaes/apps/uncertainty_propagation/sens.py

```diff
@@ -0,0 +1,731 @@
+        Provides options to the solver (also the name of an attribute)
+    var_dic: dictionary
+        If any original variable contains "'", need an auxiliary dictionary 
+        with keys theta_names without "'", values with "'".
+        e.g) var_dic: {'fs.properties.tau[benzene,toluene]': "fs.properties.tau['benzene','toluene']",
+                       'fs.properties.tau[toluene,benzene]': "fs.properties.tau['toluene','benzene']"}
+
+    Returns
+    -------
+    dsdp: numpy.ndarray
+        Gradient vector of the (decision variables, parameters) with respect to paramerters (=theta_name).
+        number of rows = len(theta_name)
+        number of columns= len(col)
+    col: list
+        List of variable names
+    """
+    m = model.clone()
+    original_Param = []
+    perturbed_Param = []
+    m.extra = ConstraintList()
+    kk = 0
+    if var_dic == {}:
+        for i in theta_names:
+            var_dic[i] = i
+    for v in theta_names:
+        v_tmp = str(kk)
+        original_param_object = Param(initialize=theta[v], mutable=True)
+        perturbed_param_object = Param(initialize=theta[v])
+        m.add_component("original_"+v_tmp, original_param_object)
+        m.add_component("perturbed_"+v_tmp, perturbed_param_object)
+        m.extra.add(eval('m.'+var_dic[v]) - eval('m.original_'+v_tmp) == 0 )
+        original_Param.append(original_param_object)
+        perturbed_Param.append(perturbed_param_object)
+        kk = kk + 1
+    m_kaug_dsdp = sensitivity_calculation('kaug',m,original_Param,perturbed_Param, tee)
+
+    try:
+        with open ("./dsdp/col_row.col", "r") as myfile:
+            col = myfile.read().splitlines()
+        dsdp = np.loadtxt("./dsdp/dsdp_in_.in")
+    except Exception as e:
+        print('File not found.')
+    dsdp = dsdp.reshape((len(theta_names), int(len(dsdp)/len(theta_names))))
+    dsdp = dsdp[:len(theta_names), :len(col)]
+    '''
+    dsdp_dic = {}
+    for i in range(len(theta_names)):
+        for j in range(len(col)):
+            if SensitivityInterface.get_default_block_name() not in col[j]:
+                dsdp_dic["d("+col[j] +")/d("+theta_names[i]+")"] =  -dsdp[i, j]
+    '''
+    try:
+        shutil.rmtree('dsdp', ignore_errors=True)
+    except OSError:
+        pass
+    col = [i for i in col if SensitivityInterface.get_default_block_name() not in i]
+    return dsdp, col
+
+def get_dfds_dcds(model, theta_names, tee=False, solver_options=None):
+    """This function calculates gradient vector of the objective function 
+       and constraints with respect to the variables in theta_names.
+    e.g) min f:  p1*x1+ p2*(x2^2) + p1*p2
+         s.t  c1: x1 + x2 = p1
+              c2: x2 + x3 = p2
+              0 <= x1, x2, x3 <= 10
+              p1 = 10
+              p2 = 5
+    - Variables = (x1, x2, x3, p1, p2)
+    - Fix p1 and p2 with estimated values
+    - The function provides gradient vector at the optimal solution
+      gradient vector of the objective function, 
```

Comment:
Update this to reflect you are returning sparse numpy arrays?

Reply:
Added dimension. But with numpy array.